common/hashing.h

```diff
@@ -474,6 +481,63 @@ inline auto CarbonHashValue(const llvm::SmallString<Length>& value,
   return CarbonHashValue(llvm::StringRef(value.data(), value.size()), seed);
 }
 
+// Support types that are array like by building an `llvm::ArrayRef` out of
+// them. We can't do this by accepting any type convertible to an `ArrayRef`
+// because that type supports building a synthetic array out of any single
+// element. These need to go back through the dispatch implementation though to
+// pick up the full space of hash overloads for `T`.
+template <typename T>
+inline auto CarbonHashValue(const std::vector<T>& arg, uint64_t seed)
+    -> HashCode {
+  return DispatchImpl(llvm::ArrayRef(arg), seed);
+}
+template <typename T>
+inline auto CarbonHashValue(const llvm::SmallVectorImpl<T>& arg, uint64_t seed)
+    -> HashCode {
+  return DispatchImpl(llvm::ArrayRef(arg), seed);
+}
+template <typename T, size_t N>
+inline auto CarbonHashValue(const std::array<T, N>& arg, uint64_t seed)
+    -> HashCode {
+  return DispatchImpl(llvm::ArrayRef(arg), seed);
+}
+template <typename T, size_t N>
+inline auto CarbonHashValue(const T (&arg)[N], uint64_t seed) -> HashCode {
+  return DispatchImpl(llvm::ArrayRef(arg), seed);
+}
+
+inline auto CarbonHashValue(llvm::APInt value, uint64_t seed) -> HashCode {
+  Hasher hasher(seed);
+  if (LLVM_LIKELY(value.isSingleWord())) {
+    hasher.Hash(value.getBitWidth(), value.getZExtValue());
+  } else {
+    hasher.Hash(value.getBitWidth());
+    hasher.HashSizedBytes(
+        llvm::ArrayRef(value.getRawData(), value.getNumWords()));
+  }
+  return static_cast<HashCode>(hasher);
+}
+
+inline auto CarbonHashValue(llvm::APFloat value, uint64_t seed) -> HashCode {
+  Hasher hasher(seed);
+  // Hashing floating point numbers is complex and depends on the specific
+  // internal semantics of `APFloat`, so delegate to the LLVM hashing framework
+  // here. We re-hash the result to mix in our seed. All of this is a bit
+  // inefficient, and we can revisit this to provide a dedicated implementation
+  // if it becomes a bottleneck.
+  using llvm::hash_value;
+  hasher.Hash(hash_value(value));
+  return static_cast<HashCode>(hasher);
+}
+
+// Implementation detail predicate to test for types hashable with
+// `CarbonHashValue`. Should find all of the above direct overloads as well as
+// ADL-provided overloads.
+template <typename T>
+concept HasCustomCarbonHashValue = requires(T t) {
+  { CarbonHashValue(t, static_cast<uint64_t>(0)) } -> std::same_as<HashCode>;
+};
```

Comment:
I'm nervous about depending on this finding the overloads declared above this line but not the ones declared below -- that seems a bit subtle, as reordering the code would result in it compiling but meaning something different. (It also wouldn't work on MSVC...)

Would it be feasible to give a different name to the non-custom overloads below, so we go through two layers of customization, first for raw hashing then for custom hashing?